src/hotspot/share/runtime/os.cpp

```diff
@@ -2023,3 +2030,23 @@ void os::PageSizes::print_on(outputStream* st) const {
     st->print("empty");
   }
 }
+
+void os::record_time_before_checkpoint() {
+  // If CRaC supports multiple checkpoint - restore - checkpoint - restore cycles
+  // we want to record the timestamps only on the first checkpoint, but update
+  // the offset after each restore
+  if (checkpoint_millis < 0) {
+      checkpoint_millis = javaTimeMillis();
+      checkpoint_nanos = javaTimeNanos();
+  }
+}
+
+void os::update_javaTimeNanos_offset() {
+  assert(checkpoint_millis >= 0, "Restore without a checkpoint?");
+  long diff_millis = javaTimeMillis() - checkpoint_millis;
+  // If the wall clock has gone backwards we won't add it to the offset
+  if (diff_millis < 0) {
+    diff_millis = 0;
+  }
+  javaTimeNanos_offset = checkpoint_nanos - javaTimeNanos() + diff_millis * 1000000L;
```

Comment:
Can you please explain why isn't `javaTimeNanos_offset = checkpoint_nanos - javaTimeNanos()` sufficient? What am I  missing here?

Reply:
@ashu-mehra Had we ignored the `diff_millis` part, the diff of nanotime before and after checkpoint would indicate that no time elapsed between checkpoint and restore. With this in place the difference does not have the expected accuracy but at least it's monotonic; wall clock time difference gives us the best possible estimate.